src/kernel/thread/thread_cancel_enable.c

```diff
@@ -13,6 +13,19 @@
 #include <kernel/thread/thread_cancel.h>
 
 int thread_cancel(struct thread *t) {
+	if (t->state == TS_EXITED) {
+		return ESRCH;
+	}
+	if (PTHREAD_CANCEL_DISABLE == t->cleanups.state) {
+		while (1) {
+			if (PTHREAD_CANCEL_ENABLE == t->cleanups.state) {
+				break;
+			}
+			if (t->state == TS_EXITED) {
+				return 0;
+			}
+		}
```

Comment:
What are you doing here? Are you waiting for the thread state to be exited? Did you investigate where this state will be changed?

Reply:
You are waiting for PTHREAD_CANCEL_ENABLE, but where it will be set?

Reply:
> What are you doing here? Are you waiting for the thread state to be exited? Did you investigate where this state will be changed?

Yes. It is waiting for either the state change or the completion of the thread. If the state changes from `PTHREAD_CANCEL_DISABLE` to  `PTHREAD_CANCEL_ENABLE` then thread cancelation can be done and if the thread executes without changing the state (`PTHREAD_CANCEL_DISABLE`) then we can not apply `pthread_cancel` on it.

> You are waiting for PTHREAD_CANCEL_ENABLE, but where it will be set?

I have added a test for this situation where a thread disables its cancelation then executes some task then again enables its cancelation. [Here](https://github.com/sobhaniiest/embox/blob/da704f750c7d0649c6779f53efe2f7fb5a8e4e3d/src/tests/posix/pthread/pthread_cancel_test.c#L83), please have a look.

I've got this example from [here](http://man7.org/linux/man-pages/man3/pthread_cancel.3.html).